### ISSUE TITLE
AOTDispatcher debug mode

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#136272]
(from discussion with @bdhirsh and @ezyang)

Motivation: AOTDispatcher (and it's interaction with HOPs) is generally difficult to understand. The claim is that the complexity is around the PyTorch dispatcher applying successive transforms: we apply Autograd, subclass handling, Functionalization, in that order, but "all at the same time": it materializes one FX graph where these transforms have been applied. Developers who work with AOTDispatcher need to understand the sequence of dispatch and keep it around in their heads.

Pitch: Instead, we should offer a debug mode that applies each transform to a graph. For example:
- we would trace out a joint graph that operates on Tensor subclasses
- then we would re-trace that graph to produce another joint graph that doesn't involve Tensor subclasses
- next we would re-trace the previous graph to produce a joint graph that is functionalized.
- finally we would apply the decomposition table.

The presence of the intermediate graphs lets developers see how the successive transforms work improve debuggability. An analogy is that in TORCH_LOGS, we let people see the FX graph at different stages of torch.compile; it would be difficult to debug if the only graph we materialized was Inductor's output_code.

**Comment by bdhirsh** [comment]
cc @drisspg (think this would help people?)

**Comment by drisspg** [comment]
yeah I could see this being useful I wonder what a good example of a bug/issue where if we had this feature it would have been easier to reason about
